File: Weather_App/WeatherApp.API/Services/WeatherService.cs
using WeatherApp.API.Models;

namespace WeatherApp.API.Services
{
    public interface IWeatherService
    {
        Task<WeatherData> GetCurrentWeatherAsync(string city);
        Task<WeatherForecast> GetWeatherForecastAsync(string city);
        Task<List<string>> GetPopularCitiesAsync();
    }

    public class WeatherService : IWeatherService
    {
        private readonly IOpenWeatherService _openWeatherService;
        private readonly ILogger<WeatherService> _logger;
        private readonly IConfiguration _configuration;

        public WeatherService(IOpenWeatherService openWeatherService, ILogger<WeatherService> logger, IConfiguration configuration)
        {
            _openWeatherService = openWeatherService;
            _logger = logger;
            _configuration = configuration;
        }

        public async Task<WeatherData> GetCurrentWeatherAsync(string city)
        {
            try
            {
                var openWeatherData = await _openWeatherService.GetCurrentWeatherAsync(city);
                
                return new WeatherData
                {
                    City = openWeatherData.Name,
                    Temperature = Math.Round(openWeatherData.Main.Temp, 1),
                    FeelsLike = Math.Round(openWeatherData.Main.FeelsLike, 1),
                    Humidity = openWeatherData.Main.Humidity,
                    Description = openWeatherData.Weather.FirstOrDefault()?.Description ?? "Unknown",
                    Icon = openWeatherData.Weather.FirstOrDefault()?.Icon ?? "01d",
                    WindSpeed = Math.Round(openWeatherData.Wind.Speed, 1),
                    Timestamp = DateTimeOffset.FromUnixTimeSeconds(openWeatherData.Timestamp).UtcDateTime
                };
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error getting current weather for {City}", city);
                throw;
            }
        }

        public async Task<WeatherForecast> GetWeatherForecastAsync(string city)
        {
            try
            {
                var openWeatherForecast = await _openWeatherService.GetForecastAsync(city);
                var forecast = new WeatherForecast
                {
                    City = openWeatherForecast.City.Name,
                    threeHourForecast = new List<WeatherData>()
                };
                
                _logger.LogInformation("Processing forecast for {City}, {Count} items", city, openWeatherForecast.List.Count);
                
                // Take up to 40 items (5 days, 3-hour intervals)
                foreach (var item in openWeatherForecast.List.Take(40))
                {
                    var timestamp = DateTimeOffset.FromUnixTimeSeconds(item.Timestamp).UtcDateTime;
                    forecast.threeHourForecast.Add(new WeatherData
                    {
                        City = openWeatherForecast.City.Name,
                        Temperature = Math.Round(item.Main.Temp, 1),
                        FeelsLike = Math.Round(item.Main.FeelsLike, 1),
                        Humidity = item.Main.Humidity,
                        Description = item.Weather.FirstOrDefault()?.Description ?? "Unknown",
                        Icon = item.Weather.FirstOrDefault()?.Icon ?? "01d",
                        WindSpeed = Math.Round(item.Wind.Speed, 1),
                        Timestamp = timestamp
                    });
                }
                
                _logger.LogInformation("Returning {Count} forecast items for {City}", forecast.threeHourForecast.Count, city);
                return forecast;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error getting forecast for {City}", city);
                throw;
            }
        }

        public Task<List<string>> GetPopularCitiesAsync()
        {
            // Get popular cities from configuration
            var popularCities = _configuration.GetSection("PopularCities").Get<List<string>>();

            // Return the list, or an empty list if not found in configuration
            return Task.FromResult(popularCities ?? new List<string>());
        }
    }
} 

File: Weather_App/src/app/services/weather.service.ts
import { Injectable, Injector, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface WeatherData {
  city: string;
  temperature: number;
  feelsLike: number;
  humidity: number;
  description: string;
  icon: string;
  windSpeed: number;
  timestamp: string;
}

export interface WeatherForecast {
  city: string;
  threeHourForecast: WeatherData[];
}

export interface WeatherRequest {
  city: string;
}

@Injectable({
  providedIn: 'root'
})
export class WeatherService {
  private apiUrl = 'http://localhost:5000/api/weather';

  constructor(private http: HttpClient) {}


  getCurrentWeather(city: string): Observable<WeatherData> {
    const encodedCity = encodeURIComponent(city);
    return this.http.get<WeatherData>(`${this.apiUrl}/current/${encodedCity}`);
  }

  getWeatherForecast(city: string): Observable<WeatherForecast> {
    const encodedCity = encodeURIComponent(city);
    return this.http.get<WeatherForecast>(`${this.apiUrl}/forecast/${encodedCity}`);
  }

  getPopularCities(): Observable<string[]> {
    return this.http.get<string[]>(`${this.apiUrl}/cities`);
  }

  searchWeather(request: WeatherRequest): Observable<WeatherData> {
    return this.http.post<WeatherData>(`${this.apiUrl}/search`, request);
  }
} 